### PR TITLE
Intitial release

### PR DESCRIPTION
**Pull request by derklaro**
This is the initial version of the communication protocol. There are some last things missing, mainly documentation. This will be added later today.

### CODE REVIEW

**Review by zLvkas — COMMENTED** [review]
Everything is perfect, just like everything else about you.

**Review by derrop — APPROVED** [review]

**Review by zLvkas — APPROVED** [review]
ok

File: core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                channelHandlerContext.channel().close().syncUninterruptibly();
+                return;
+            }
+
+            list.add(packet);
+        } catch (Throwable throwable) {
+            throwable.printStackTrace();
+            channelHandlerContext.channel().close().syncUninterruptibly();
+        } finally {
+            // TODO: ByteBufUtil.releaseFully(byteBuf);
```
Comment (derklaro) [inline core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java:83]: @derrop what was the issue here?

File: core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            byteBuf.writeBytes(fileData);
+        } catch (IOException exception) {
+            exception.printStackTrace();
+        }
+    }
+
+    public static void readFile(@NotNull Path path, @NotNull ByteBuf byteBuf) {
+        int length = byteBuf.readInt();
+        try (OutputStream outputStream = Files.newOutputStream(path)) {
+            byteBuf.readBytes(outputStream, length); // TODO: not tested
```
Comment (derklaro) [inline core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java:64]: @derrop did you test this?

File: core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java (hunk 1 above)
Comment (derrop) [reply core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java:83]: threw an IllegalRefCntException

File: core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java (hunk 2 above)
Comment (derrop) [reply core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java:64]: no

File: core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java (hunk 1 above)
Comment (derklaro) [reply core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java:83]: Can you submit me the exception?

File: core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java (hunk 2 above)
Comment (derklaro) [reply core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java:64]: Can you test it?

File: core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java (hunk 1 above)
Comment (derrop) [reply core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java:83]: Just start a server and a client, I don't have the logs anymore.

Comment (derklaro) [reply core/src/main/java/de/hdskins/protocol/netty/packet/PacketDecoder.java]: Resolved

File: core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java (hunk 2 above)
Comment (derklaro) [reply core/src/main/java/de/hdskins/protocol/netty/ByteBufUtil.java]: Tested it and reformatted the code, resolved

**Pull request merged**